File: src/libs/m3/cap/Gate.cc
/*
 * Copyright (C) 2016, Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <m3/cap/Gate.h>
#include <m3/Syscalls.h>
#include <m3/Errors.h>

namespace m3 {

Errors::Code Gate::async_cmd(Operation op, void *data, size_t datalen, size_t off, size_t size,
        label_t reply_lbl, int reply_ep) {
    // ensure that the DMAUnit is ready. this is required if we want to mix async sends with
    // sync sends.
    wait_until_sent();
    ensure_activated();

retry:
    Errors::Code res;
    switch(op) {
        case SEND:
            res = DTU::get().send(_epid, data, datalen, reply_lbl, reply_ep);
            break;
        case READ:
            res = DTU::get().read(_epid, data, datalen, off);
            break;
        case WRITE:
            res = DTU::get().write(_epid, data, datalen, off);
            break;
        case CMPXCHG:
            res = DTU::get().cmpxchg(_epid, data, datalen, off, size);
            break;
    }
    if(res == Errors::VPE_GONE) {
        res = Syscalls::get().activate(_epid, sel(), sel());
        if(res != Errors::NO_ERROR)
            return res;
        goto retry;
    }
    return res;
}

}

File: src/libs/rustm3/src/com/gate.rs
/*
 * Copyright (C) 2018, Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

use cap::{CapFlags, Capability, Selector};
use cell::Cell;
use com::EpMux;
use core::ops;
use dtu::EpId;
use errors::Error;

#[derive(Debug)]
pub struct Gate {
    cap: Capability,
    ep: Cell<Option<EpId>>,
}

impl Gate {
    pub fn new(sel: Selector, flags: CapFlags) -> Self {
        Self::new_with_ep(sel, flags, None)
    }

    pub const fn new_with_ep(sel: Selector, flags: CapFlags, ep: Option<EpId>) -> Self {
        Gate {
            cap: Capability::new(sel, flags),
            ep: Cell::new(ep),
        }
    }

    pub fn sel(&self) -> Selector {
        self.cap.sel()
    }

    pub fn flags(&self) -> CapFlags {
        self.cap.flags()
    }

    pub fn ep(&self) -> Option<EpId> {
        self.ep.get()
    }
    pub fn set_ep(&self, ep: EpId) {
        self.ep.set(Some(ep))
    }
    pub fn unset_ep(&self) {
        self.ep.set(None)
    }

    pub fn activate(&self) -> Result<EpId, Error> {
        match self.ep() {
            Some(ep) => Ok(ep),
            None     => EpMux::get().switch_to(self),
        }
    }

    pub fn rebind(&mut self, sel: Selector) -> Result<(), Error> {
        EpMux::get().switch_cap(self, sel)?;
        self.cap.rebind(sel);
        Ok(())
    }
}

impl ops::Drop for Gate {
    fn drop(&mut self) {
        EpMux::get().remove(self);
    }
}
